src/webnn/native/dml/GraphDML.cpp

```diff
@@ -1675,7 +1701,289 @@ namespace webnn::native ::dml {
+        if (!CreateDmlTensorDesc(mDmlTensorsDesc, weightDmlTensorDesc, &constantWeightTensorDesc,
+                                 {}, {}, true)) {
+            return DAWN_INTERNAL_ERROR("Failed to create DML tensor description.");
+        }
+        DML_TENSOR_DESC weightTensorDesc = {DML_TENSOR_TYPE_BUFFER,
+                                            &weightDmlTensorDesc->bufferDesc};
+        { APPEND_IDENTITY(constantWeightTensorDesc, weightTensorDesc, dmlOperator); }
+        auto weightEdge = CreateEdgeFromThisNode(weightTensorDesc, mGraphDesc.NodeCount());
+        AddNodeAndEdgesToGraphDesc(mGraphDesc, {constantWeightEdge}, dmlOperator);
+        inputEdges.push_back(weightEdge);
+
+        // Recurrence: 4D tensor with the Sizes { 1, num_directions, 3 * hidden_size, hidden_size }.
+        // Need to reshape recurrence from WebNN 3-D to DML 4-D.
+        // Need to convert tensor flag to NONE.
+        auto constantRecurrenceEdge = mGraphEdgesMap[inputsOperand[2].Get()];
+        auto webnnRecurrenceDims = ConvertDimensions(inputsOperand[2].Get()->Shape());
+        std::vector<UINT> recurrenceDims = {1, webnnRecurrenceDims[0], webnnRecurrenceDims[1],
+                                            webnnRecurrenceDims[2]};
+        std::shared_ptr<DmlTensorDesc> constantRecurrenceDmlTensorDesc(new DmlTensorDesc);
+        if (!CreateDmlTensorDesc(mDmlTensorsDesc, constantRecurrenceDmlTensorDesc,
+                                 &constantRecurrenceEdge->outputTensorDESC, recurrenceDims)) {
+            return DAWN_INTERNAL_ERROR("Failed to create DML tensor description.");
+        }
+        DML_TENSOR_DESC constantRecurrenceTensorDesc = {
+            DML_TENSOR_TYPE_BUFFER, &constantRecurrenceDmlTensorDesc->bufferDesc};
+        std::shared_ptr<DmlTensorDesc> recurrenceDmlTensorDesc(new DmlTensorDesc);
+        if (!CreateDmlTensorDesc(mDmlTensorsDesc, recurrenceDmlTensorDesc,
+                                 &constantRecurrenceTensorDesc, {}, {}, true)) {
+            return DAWN_INTERNAL_ERROR("Failed to create DML tensor description.");
+        }
+        DML_TENSOR_DESC recurrenceTensorDesc = {DML_TENSOR_TYPE_BUFFER,
+                                                &recurrenceDmlTensorDesc->bufferDesc};
+        { APPEND_IDENTITY(constantRecurrenceTensorDesc, recurrenceTensorDesc, dmlOperator); }
+        auto recurrenceEdge = CreateEdgeFromThisNode(recurrenceTensorDesc, mGraphDesc.NodeCount());
+        AddNodeAndEdgesToGraphDesc(mGraphDesc, {constantRecurrenceEdge}, dmlOperator);
+        inputEdges.push_back(recurrenceEdge);
+
+        const GruOptions* options = gru->GetOptions();
+        UINT operandIndex = 3;
+
+        // Bias: 4D tensor with the Sizes of { 1, 1, num_directions, 6 * hidden_size }.
+        // Need to concat bias tensor and recurrentBias tensor.
+        // Need to reshape bias from WebNN 2-D to DML 4-D.
+        std::vector<UINT> webnnBiasDims = {weightDims[1],
```

Comment:
Can we get shape for all webnn dims?

Reply:
Yes, but here we can also infer from other inputs, because even if the bias parameter is not passed in, we still need to use this.

Reply:
OK，got it, thanks! @miaobin